### PR TITLE
Keras 3.3.3 upgrade

### PR DESCRIPTION
**Pull request by Chase-Grajeda**
- Included Keras 3.3.3 upgrades for all network classes and instances of ``tf.model``, ``tf.layer``, and ``tf.optimizer``
- Added ``@tf.function`` decorator to ``coupling_networks._calculate_spline()`` to allow execution in non-eager environments
- Changed instances of ``tf.Tensor`` in docstring to ``Tensor`` w.r.t. the experimental Tensor type

The included changes have worked successfully in all of the example / tutorial notebooks, however I'm sure there are a few introduced bugs laying around when TensorFlow is not the chosen backend. 

Current environment dependencies:
- keras 3.3.3
- tensorflow 2.16.1
- tensorflow-intel 2.16.1
- tensorflow-probability 0.24.0

### CODE REVIEW

File: bayesflow/coupling_networks.py
Comment (LarsKue) [inline bayesflow/coupling_networks.py:442]: tensorflow-specific, `len(shape)` or `parameters.ndim` should be fine. Safest would probably be `keras.ops.ndim`

File: bayesflow/helper_functions.py
Comment (LarsKue) [inline bayesflow/helper_functions.py:115]: not sure if this works under jax/numpy backend - can you test this? As an alternative, we could use `keras.ops.convert_to_numpy()`. We can then drop the `hasattr` check

File: bayesflow/inference_networks.py
Comment (LarsKue) [inline bayesflow/inference_networks.py:193]: as discussed, please use `keras.ops`

**Review by LarsKue — CHANGES_REQUESTED** [review]
Very nice changes overall, thank you! See comments for some minor changes before merging.

File: bayesflow/coupling_networks.py
Comment (Chase-Grajeda) [reply bayesflow/coupling_networks.py]: Changed `parameters.shape.rank` to use `keras.ops.ndim` in coupling_networks.py (see [47fbbbd](https://github.com/stefanradev93/BayesFlow/pull/165/commits/47fbbbd4acb22c391b483c845f73e6862094c07a))

File: bayesflow/helper_functions.py
Comment (Chase-Grajeda) [reply bayesflow/helper_functions.py:115]: Changed `learning_rate().numpy()` to `keras.ops.convert_to_numpy()` in helper_functions.py (see [47fbbbd](https://github.com/stefanradev93/BayesFlow/pull/165/commits/47fbbbd4acb22c391b483c845f73e6862094c07a))

File: bayesflow/inference_networks.py
Comment (Chase-Grajeda) [reply bayesflow/inference_networks.py]: Changed requested `keras.backend` instances to `keras.ops` in `inference_networks.py`. (see [47fbbbd](https://github.com/stefanradev93/BayesFlow/pull/165/commits/47fbbbd4acb22c391b483c845f73e6862094c07a))
Will change remaining `keras.backend` to `keras.ops` calls with next update

Comment (Chase-Grajeda) [reply bayesflow/inference_networks.py]: Remaining `keras.backend` calls switched to respective `keras.ops` functions (see [b9d5f16](https://github.com/stefanradev93/BayesFlow/pull/165/commits/b9d5f16a9e2a9b80db03f7d8262a4b39d15e2f8c))

**Comment by LarsKue** [comment]
We cannot merge this anymore due to the many changes on the branch, but we'll use this as a reference nonetheless.

**Pull request closed without merging**
